### PR TITLE
Enforce permissions

### PR DESCRIPTION
**Pull request by argerus**
Parsing JWT access tokens and making them available as `Permissions` in the GRPC handling code of databroker was previously implemented in #501.

This PR adds enforcement of these permissions throughout databroker.

This is accomplished by introducing two new interfaces `DatabaseReadAccess` and `DatabaseWriteAccess` that can only be created by supplying a `Permissions` struct.

All functions that read / write or create entries can only be accessed through these interfaces (structs) and that's where the enforcement happens.

In addition, `authorized_access(Permissions) -> AuthorizedAccess` has been added to the top level `DataBroker` interface in order to provide a convenient way to set the permissions once, and use the wrapper functions (in `AuthorizedAccess`) to interact with the previously mentioned interfaces.

### CODE REVIEW

**Review by lukasmittag — COMMENTED** [review]
Error output varies (only loveliness issue): ```Error setting id 51: AccessDenied``` ```ehicle.Body.Lights.IsRunningOn: ( AccessDenied )``` ```Error setting Vehicle.Body.Trunk.Front.IsOpen: AccessDenied```

File: kuksa_databroker/databroker/src/broker.rs [hunk 1]
```diff
@@ -34,6 +35,15 @@ pub enum UpdateError {
...
+    PermissionDenied,
+    PermissionExpired,
+}
+
+#[derive(Debug)]
+pub enum ReadError {
+    NotFound,
+    PermissionDenied,
+    PermissionExpired,
 }
```
[the remark this answers is not in the document]
Comment (argerus) [reply kuksa_databroker/databroker/src/broker.rs:48]: Good point. And yes, I agree.
I postponed the `create` scope and permission handling for some reason, but I really should add that as well.

File: kuksa_databroker/databroker/src/jwt/decoder.rs [hunk 2]
```diff
@@ -87,60 +87,50 @@ impl Decoder {
...
-                            actuate.extend_with(permissions::PathMatcher::Everything)
+                            permissions.add_actuate_permission(Permission::All)
                         }
                         scope::Action::Provide => {
-                            provide.extend_with(permissions::PathMatcher::Everything)
+                            permissions.add_provide_permission(Permission::All)
                         }
-                    }
+                    };
                 }
```
[the remark this answers is not in the document]
Comment (argerus) [reply kuksa_databroker/databroker/src/jwt/decoder.rs:129]: Added `create` scope as well.

File: kuksa_databroker/databroker/src/broker.rs [hunk 3]
```diff
@@ -782,69 +886,98 @@ impl DataBroker {
         entry_type: EntryType,
         description: String,
     ) -> Result<i32, RegistrationError> {
-        self.entries
+        self.broker
+            .database
             .write()
             .await
+            .authorized_write_access(self.permissions)
```
[the remark this answers is not in the document]
Comment (argerus) [reply kuksa_databroker/databroker/src/broker.rs:901]: This isn't an actual check, it's just how a set of `permissions` are "bound" to the following set of interactions that happens through the returned `DatabaseWriteAccess`.

So, you are right, the real check happens in `DatabaseWriteAccess::add(...)`. Or as you point out, should happen.

It has been added now.

**Comment by argerus** [comment]
> Error output varies (only loveliness issue): Error setting id 51: AccessDenied ehicle.Body.Lights.IsRunningOn: ( AccessDenied ) Error setting Vehicle.Body.Trunk.Front.IsOpen: AccessDenied

That's an artifact of the fact that the `Collector` and `Broker` interfaces use `path` and `i32` respectively to refer to the entries / datapoints / signals.

Anyway, I changed the error messages, since it's a lot more user friendly to refer to them by path.

File: kuksa_databroker/databroker/src/permissions.rs [hunk 4]
```diff
@@ -13,56 +13,207 @@
...
+        self.expired()?;
+
+        if self.create.is_match(path) {
+            return Ok(());
+        }
+        Err(PermissionError::Denied)
+    }
+
+    #[inline]
+    pub fn expired(&self) -> Result<(), PermissionError> {
```
Comment (lukasmittag) [inline kuksa_databroker/databroker/src/permissions.rs:210]: shouldn't this function be above the first use of it? For reading and understanding purposes.

File: kuksa_databroker/databroker-examples/examples/perf_setter.rs [hunk 5]
```diff
@@ -64,6 +71,9 @@ async fn run_streaming_set_test(iterations: i32, n_th_message: i32) {
                 }
             };
 
+            if datapoint1_id == -1 {
+                return;
+            }
```
Comment (lukasmittag) [inline kuksa_databroker/databroker-examples/examples/perf_setter.rs:76]: Couldn't this be handeled in line 70?

**Review by lukasmittag — APPROVED** [review]

**Comment by lukasmittag** [comment]
Just did some minor comments. Nothing that is necessary so approved it. @SebastianSchildt now it's your turn :)

**Pull request merged**